models/completed_uploads.go

```diff
@@ -0,0 +1,202 @@
+	), nil
+}
+
+// ValidateCreate gets run every time you call "pop.ValidateAndCreate" method.
+// This method is not required and may be deleted.
+func (c *CompletedUpload) ValidateCreate(tx *pop.Connection) (*validate.Errors, error) {
+	return validate.NewErrors(), nil
+}
+
+// ValidateUpdate gets run every time you call "pop.ValidateAndUpdate" method.
+// This method is not required and may be deleted.
+func (c *CompletedUpload) ValidateUpdate(tx *pop.Connection) (*validate.Errors, error) {
+	return validate.NewErrors(), nil
+}
+
+/**
+ * Callbacks
+ */
+
+func (c *CompletedUpload) BeforeCreate(tx *pop.Connection) error {
+
+	// Defaults to ClaimNotBegun
+	if c.PRLStatus == 0 {
+		c.PRLStatus = PRLClaimNotStarted
+	}
+
+	// Defaults to GasTransferNotStarted
+	if c.GasStatus == 0 {
+		c.GasStatus = GasTransferNotStarted
+	}
+	return nil
+}
+
+/**
+ * Methods
+ */
+func NewCompletedUpload(session UploadSession) error {
+
+	var err error
+
+	switch session.Type {
+	case SessionTypeAlpha:
+		_, err = DB.ValidateAndSave(&CompletedUpload{
+			GenesisHash:   session.GenesisHash,
+			ETHAddr:       session.ETHAddrAlpha.String,
+			ETHPrivateKey: session.ETHPrivateKey})
+	case SessionTypeBeta:
+		_, err = DB.ValidateAndSave(&CompletedUpload{
+			GenesisHash:   session.GenesisHash,
+			ETHAddr:       session.ETHAddrBeta.String,
+			ETHPrivateKey: session.ETHPrivateKey})
+	default:
+		err = errors.New("no session type provided for session in method models.NewCompletedUpload")
+	}
+
+	return err
+}
+
+func GetRowsByGasAndPRLStatus(gasStatus int, prlStatus int) (uploads []CompletedUpload, err error) {
+	err = DB.Where("gas_status = ? AND prl_status = ?", gasStatus, prlStatus).All(&uploads)
+	return uploads, err
+}
+
+func GetRowsByGasStatus(gasStatus int) (uploads []CompletedUpload, err error) {
+	err = DB.Where("gas_status = ?", gasStatus).All(&uploads)
+	return uploads, err
+}
+
+func SetGasStatus(uploads []CompletedUpload, newGasStatus int) {
+	for _, upload := range uploads {
+		upload.GasStatus = newGasStatus
+		DB.ValidateAndSave(&upload)
+	}
+}
+
+func GetRowsByPRLStatus(prlStatus int) (uploads []CompletedUpload, err error) {
+	err = DB.Where("prl_status = ?", prlStatus).All(&uploads)
+	return uploads, err
+}
+
+func SetPRLStatus(uploads []CompletedUpload, newPRLStatus int) {
+	for _, upload := range uploads {
+		upload.PRLStatus = newPRLStatus
+		DB.ValidateAndSave(&upload)
+	}
+}
+
+func GetTimedOutGasTransfers(thresholdTime time.Time) (uploads []CompletedUpload, err error) {
+	err = DB.Where("gas_status = ? AND updated_at <= ?",
+		GasTransferProcessing,
+		thresholdTime).All(&uploads)
+	return uploads, err
+}
+
+func GetTimedOutPRLTransfers(thresholdTime time.Time) (uploads []CompletedUpload, err error) {
+	err = DB.Where("prl_status = ? AND updated_at <= ?",
+		PRLClaimProcessing,
+		thresholdTime).All(&uploads)
+	return uploads, err
+}
+
+func SetGasStatusByAddress(transactionAddress string, newGasStatus int) {
```

Comment:
The intention with this method and the one below is for the eth_gateway file to all this when it hears a successful transfer event, to change the GasStatus or PRLStatus

Reply:
Instead of this API taking in an int, we can add a type to constrain it:

```go
type GasTransferStatus int
const (
	GasTransferNotStarted GasTransferStatus = iota + 1
	GasTransferProcessing
	GasTransferSuccess
	GasTransferError = -1
)
// ...
func SetGasStatusByAddress(transactionAddress string, status GasTransferStatus) 
```

Reply:
sure